// react-atlas-core/lib/tabs/Tabs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _reactThemeable = require("react-themeable");

var _reactThemeable2 = _interopRequireDefault(_reactThemeable);

var _utils = require("../utils");

var _Tab = require("./Tab");

var _Tab2 = _interopRequireDefault(_Tab);

var _TabContent = require("./TabContent");

var _TabContent2 = _interopRequireDefault(_TabContent);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Wrapper component to organize and produce tabs using multiple `<Tab>` components as children.
 */
var Tabs = function (_Component) {
  _inherits(Tabs, _Component);

  function Tabs() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Tabs);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Tabs.__proto__ || Object.getPrototypeOf(Tabs)).call.apply(_ref, [this].concat(args))), _this), _this._handleHeaderClick = function (idx) {
      if (_this.props.onChange) {
        _this.props.onChange(idx);
      }
    }, _this._parseChildren = function () {
      var headers = [];
      var contents = [];

      _react2.default.Children.forEach(_this.props.children, function (item) {
        if (item.type === _Tab2.default) {
          headers.push(item);
          if (item.props.children) {
            contents.push(_react2.default.createElement(_TabContent2.default, {
              children: item.props.children,
              theme: _this.props.theme
            }));
          }
        } else if (item.type === _TabContent2.default) {
          contents.push(item);
        }
      });

      return { headers: headers, contents: contents };
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Tabs, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _parseChildren = this._parseChildren(),
          headers = _parseChildren.headers,
          contents = _parseChildren.contents;

      var theme = (0, _reactThemeable2.default)(this.props.theme);
      var classes = (0, _utils.classNames)({
        "container": true
      }, this.props.className);

      return _react2.default.createElement(
        "div",
        theme.apply(undefined, [1].concat(_toConsumableArray(classes))),
        _react2.default.createElement(
          "nav",
          _extends({}, theme(2, "navigation"), {
            ref: function ref(nav) {
              _this2.navigation = nav;
            }
          }),
          headers.map(function (item, idx) {
            return _react2.default.createElement(_Tab2.default, _extends({}, item.props, {
              key: idx,
              active: _this2.props.index === idx,
              onClick: _this2._handleHeaderClick.bind(_this2, idx),
              theme: _this2.props.theme
            }));
          })
        ),
        contents.map(function (item, idx) {
          return _react2.default.createElement(_TabContent2.default, _extends({}, item.props, {
            key: idx,
            active: _this2.props.index === idx,
            tabIndex: idx,
            theme: _this2.props.theme
          }));
        })
      );
    }
  }]);

  return Tabs;
}(_react.Component);

Tabs.defaultProps = {
  "index": 0
};

exports.default = Tabs;
