// react-atlas-default-theme/lib/table/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tr = exports.Thead = exports.Th = exports.Tfoot = exports.Td = exports.Tbody = exports.Table = undefined;

var _Table2 = require("./Table");

var _Table3 = _interopRequireDefault(_Table2);

var _Tbody2 = require("./Tbody");

var _Tbody3 = _interopRequireDefault(_Tbody2);

var _Td2 = require("./Td");

var _Td3 = _interopRequireDefault(_Td2);

var _Tfoot2 = require("./Tfoot");

var _Tfoot3 = _interopRequireDefault(_Tfoot2);

var _Th2 = require("./Th");

var _Th3 = _interopRequireDefault(_Th2);

var _Thead2 = require("./Thead");

var _Thead3 = _interopRequireDefault(_Thead2);

var _Tr2 = require("./Tr");

var _Tr3 = _interopRequireDefault(_Tr2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.Table = _Table3.default;
exports.Tbody = _Tbody3.default;
exports.Td = _Td3.default;
exports.Tfoot = _Tfoot3.default;
exports.Th = _Th3.default;
exports.Thead = _Thead3.default;
exports.Tr = _Tr3.default;
